File: backend/src/main/java/com/pm/backend/controller/TaskController.java
package com.pm.backend.controller;

import com.pm.backend.model.*;
import com.pm.backend.repository.*;
import com.pm.backend.security.UserPrincipal;
import com.pm.backend.service.AIService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/tasks")
public class TaskController {

    private final TaskRepository taskRepository;
    private final ProjectRepository projectRepository;
    private final UserRepository userRepository;
    private final CommentRepository commentRepository;
    private final NotificationRepository notificationRepository;
    private final AIService aiService;
    private final DirectMessageRepository directMessageRepository;

    public TaskController(TaskRepository taskRepository, ProjectRepository projectRepository,
                          UserRepository userRepository, CommentRepository commentRepository,
                          NotificationRepository notificationRepository, AIService aiService,
                          DirectMessageRepository directMessageRepository) {
        this.taskRepository = taskRepository;
        this.projectRepository = projectRepository;
        this.userRepository = userRepository;
        this.commentRepository = commentRepository;
        this.notificationRepository = notificationRepository;
        this.aiService = aiService;
        this.directMessageRepository = directMessageRepository;
    }

    @GetMapping("/project/{projectId}")
    public ResponseEntity<List<Task>> getProjectTasks(@PathVariable Long projectId) {
        Project project = projectRepository.findById(projectId)
                .orElseThrow(() -> new RuntimeException("Project not found: " + projectId));
        List<Task> tasks = taskRepository.findByProject(project);
        return ResponseEntity.ok(tasks);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Task> getTaskById(@PathVariable Long id) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));
        return ResponseEntity.ok(task);
    }

    @PostMapping
    public ResponseEntity<Task> createTask(@Valid @RequestBody Task task,
                                           @RequestParam(required = false) Long parentId,
                                           @AuthenticationPrincipal UserPrincipal userPrincipal) {
        User creator = userRepository.findById(userPrincipal.getId()).orElseThrow();
        task.setCreator(creator);

        Project project = projectRepository.findById(task.getProject().getId())
                .orElseThrow(() -> new RuntimeException("Project not found"));
        task.setProject(project);

        if (parentId != null) {
            Task parent = taskRepository.findById(parentId).orElseThrow();
            task.setParentTask(parent);
        }

        if (task.getStatus() == null) {
            task.setStatus("TO_DO");
        }

        // Apply AI to suggest priority if not provided
        if (task.getPriority() == null) {
            task.setPriority(aiService.predictPriority(task.getTitle(), task.getDescription(), null));
        }

        // Auto-estimate duration if empty
        if (task.getEstimatedTime() == null || task.getEstimatedTime() == 0.0) {
            int experience = task.getAssignee() != null ? 
                    userRepository.findById(task.getAssignee().getId()).map(User::getExperience).orElse(2) : 2;
            task.setEstimatedTime(aiService.estimateDuration(task.getTitle(), task.getDescription(), experience, 0));
        }

        Task savedTask = taskRepository.save(task);

        // Notify Assignee
        if (task.getAssignee() != null) {
            User assignee = userRepository.findById(task.getAssignee().getId()).orElseThrow();
            createNotification("Task Assigned", "You have been assigned: " + task.getTitle(), "TASK_ASSIGNED", assignee);

            DirectMessage dm = DirectMessage.builder()
                    .sender(creator)
                    .recipient(assignee)
                    .content("Hello, I have assigned you a new task: \"" + task.getTitle() + "\". Please review and accept it.")
                    .task(savedTask)
                    .isRead(false)
                    .build();
            directMessageRepository.save(dm);
        }

        return ResponseEntity.status(HttpStatus.CREATED).body(savedTask);
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateTask(@PathVariable Long id, @Valid @RequestBody Task taskDetails,
                                         @AuthenticationPrincipal UserPrincipal userPrincipal) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));

        User currentUser = userRepository.findById(userPrincipal.getId()).orElseThrow();
        boolean isTeamLeader = currentUser.getRole().name().equals("ROLE_ADMIN") || currentUser.getRole().name().equals("ROLE_MANAGER");
        
        // Permission enforcement: Employees can only edit tasks assigned to them
        if (!isTeamLeader) {
            if (task.getAssignee() == null || !task.getAssignee().getId().equals(currentUser.getId())) {
                System.out.println("SECURITY AUDIT WARNING: User '" + currentUser.getEmail() + "' attempted unauthorized edit on task id: " + task.getId());
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body("Only the assigned member or Team Leader can update this task.");
            }
        }

        String oldStatus = task.getStatus();
        User oldAssignee = task.getAssignee();

        task.setTitle(taskDetails.getTitle());
        task.setDescription(taskDetails.getDescription());
        task.setStatus(taskDetails.getStatus());
        task.setPriority(taskDetails.getPriority());
        task.setDueDate(taskDetails.getDueDate());
        task.setEstimatedTime(taskDetails.getEstimatedTime());
        task.setActualTime(taskDetails.getActualTime());
        task.setRecurring(taskDetails.getRecurring());
        task.setRecurringPattern(taskDetails.getRecurringPattern());

        if (taskDetails.getAssignee() != null) {
            User assignee = userRepository.findById(taskDetails.getAssignee().getId()).orElseThrow();
            task.setAssignee(assignee);
            
            // Notify if assignee has changed
            if (oldAssignee == null || !oldAssignee.getId().equals(assignee.getId())) {
                createNotification("Task Assigned", "You have been assigned: " + task.getTitle(), "TASK_ASSIGNED", assignee);

                DirectMessage dm = DirectMessage.builder()
                        .sender(currentUser)
                        .recipient(assignee)
                        .content("Hello, I have assigned you a new task: \"" + task.getTitle() + "\". Please review and accept it.")
                        .task(task)
                        .isRead(false)
                        .build();
                directMessageRepository.save(dm);
            }
        } else {
            task.setAssignee(null);
        }

        Task updatedTask = taskRepository.save(task);

        // If status transitioned from PENDING_ACCEPTANCE to TO_DO (Accept)
        if ("PENDING_ACCEPTANCE".equals(oldStatus) && "TO_DO".equals(updatedTask.getStatus()) && oldAssignee != null) {
            DirectMessage dm = DirectMessage.builder()
                    .sender(oldAssignee)
                    .recipient(task.getCreator())
                    .content("I have accepted the task: \"" + task.getTitle() + "\".")
                    .task(task)
                    .isRead(false)
                    .build();
            directMessageRepository.save(dm);
        }

        // If status transitioned from PENDING_ACCEPTANCE to BACKLOG with unassigning (Decline)
        if ("PENDING_ACCEPTANCE".equals(oldStatus) && "BACKLOG".equals(updatedTask.getStatus()) && taskDetails.getAssignee() == null && oldAssignee != null) {
            DirectMessage dm = DirectMessage.builder()
                    .sender(oldAssignee)
                    .recipient(task.getCreator())
                    .content("I have declined the task: \"" + task.getTitle() + "\". Please check the task comments for my explanation.")
                    .task(task)
                    .isRead(false)
                    .build();
            directMessageRepository.save(dm);
        }

        // Notify if task completed
        if ("COMPLETED".equals(updatedTask.getStatus()) && !"COMPLETED".equals(oldStatus)) {
            createNotification("Task Completed", "The task: " + task.getTitle() + " has been marked complete.", "TASK_COMPLETED", task.getCreator());
        }

        return ResponseEntity.ok(updatedTask);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteTask(@PathVariable Long id, @AuthenticationPrincipal UserPrincipal userPrincipal) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found: " + id));

        User currentUser = userRepository.findById(userPrincipal.getId()).orElseThrow();
        boolean isTeamLeader = currentUser.getRole().name().equals("ROLE_ADMIN") || currentUser.getRole().name().equals("ROLE_MANAGER");

        if (!isTeamLeader) {
            System.out.println("SECURITY AUDIT WARNING: User '" + currentUser.getEmail() + "' attempted unauthorized delete on task id: " + task.getId());
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body("Only Team Leaders can delete tasks.");
        }

        taskRepository.delete(task);
        return ResponseEntity.ok().build();
    }

    @GetMapping("/{id}/comments")
    public ResponseEntity<List<Comment>> getComments(@PathVariable Long id) {
        Task task = taskRepository.findById(id).orElseThrow();
        List<Comment> comments = commentRepository.findByTaskOrderByCreatedAtAsc(task);
        return ResponseEntity.ok(comments);
    }

    @PostMapping("/{id}/comments")
    public ResponseEntity<Comment> addComment(@PathVariable Long id, @RequestBody Comment comment,
                                              @AuthenticationPrincipal UserPrincipal userPrincipal) {
        Task task = taskRepository.findById(id).orElseThrow();
        User currentUser = userRepository.findById(userPrincipal.getId()).orElseThrow();
        
        comment.setTask(task);
        comment.setUser(currentUser);
        
        Comment savedComment = commentRepository.save(comment);

        // Notify creator or assignee if they are different from comment sender
        if (task.getAssignee() != null && !task.getAssignee().getId().equals(currentUser.getId())) {
            createNotification("New Comment", currentUser.getName() + " commented on: " + task.getTitle(), "MENTION", task.getAssignee());
        }

        return ResponseEntity.ok(savedComment);
    }

    @PostMapping("/{id}/timer")
    public ResponseEntity<Task> updateTimer(@PathVariable Long id, @RequestParam Double additionalHours) {
        Task task = taskRepository.findById(id).orElseThrow();
        task.setActualTime(task.getActualTime() + additionalHours);
        
        // Update project spent
        Project project = task.getProject();
        project.setSpent(project.getSpent() + (additionalHours * 50.0)); // Mock hourly rate of $50
        projectRepository.save(project);

        Task updatedTask = taskRepository.save(task);
        return ResponseEntity.ok(updatedTask);
    }

    @GetMapping("/{id}/ai-subtasks")
    public ResponseEntity<List<String>> getAiSubtasks(@PathVariable Long id) {
        Task task = taskRepository.findById(id).orElseThrow();
        List<String> suggestions = aiService.suggestSubtasks(task.getTitle(), task.getDescription());
        return ResponseEntity.ok(suggestions);
    }

    @PostMapping("/{id}/dependencies/{depId}")
    public ResponseEntity<Task> addDependency(@PathVariable Long id, @PathVariable Long depId) {
        Task task = taskRepository.findById(id).orElseThrow();
        Task dependency = taskRepository.findById(depId).orElseThrow();

        task.getDependencies().add(dependency);
        Task updatedTask = taskRepository.save(task);
        return ResponseEntity.ok(updatedTask);
    }

    @DeleteMapping("/{id}/dependencies/{depId}")
    public ResponseEntity<Task> removeDependency(@PathVariable Long id, @PathVariable Long depId) {
        Task task = taskRepository.findById(id).orElseThrow();
        Task dependency = taskRepository.findById(depId).orElseThrow();

        task.getDependencies().remove(dependency);
        Task updatedTask = taskRepository.save(task);
        return ResponseEntity.ok(updatedTask);
    }

    private void createNotification(String title, String message, String type, User recipient) {
        Notification notification = Notification.builder()
                .title(title)
                .message(message)
                .type(type)
                .recipient(recipient)
                .isRead(false)
                .build();
        notificationRepository.save(notification);
    }
}

File: backend-node/src/controllers/taskController.ts
import { Response } from 'express';
import { Task, Project, User, Comment, Notification, DirectMessage, Role } from '../models';
import { AuthRequest } from '../middleware/auth';
import { AIService } from '../services/aiService';
import { Op } from 'sequelize';

export class TaskController {
  public static async getAllTasks(_req: AuthRequest, res: Response) {
    try {
      const tasks = await Task.findAll({
        include: [
          { model: Project, as: 'project' },
          { model: User, as: 'assignee', attributes: { exclude: ['password'] } },
          { model: User, as: 'creator', attributes: { exclude: ['password'] } },
          { model: User, as: 'reviewer', attributes: { exclude: ['password'] } },
        ],
        order: [['createdAt', 'DESC']],
      });
      return res.json(tasks);
    } catch (err: any) {
      console.error('Error in getAllTasks:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async getProjectTasks(req: AuthRequest, res: Response) {
    try {
      const projectId = parseInt(req.params.projectId);
      const tasks = await Task.findAll({
        where: { projectId },
        include: [
          { model: Project, as: 'project' },
          { model: User, as: 'assignee', attributes: { exclude: ['password'] } },
          { model: User, as: 'creator', attributes: { exclude: ['password'] } },
        ],
        order: [['createdAt', 'DESC']],
      });
      return res.json(tasks);
    } catch (err: any) {
      console.error('Error in getProjectTasks:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async getTaskById(req: AuthRequest, res: Response) {
    try {
      const id = parseInt(req.params.id);
      const task = await Task.findByPk(id, {
        include: [
          { model: Project, as: 'project' },
          { model: User, as: 'assignee', attributes: { exclude: ['password'] } },
          { model: User, as: 'creator', attributes: { exclude: ['password'] } },
          { model: Task, as: 'subtasks' },
          { model: Task, as: 'dependencies', through: { attributes: [] } },
        ],
      });
      if (!task) return res.status(404).send('Task not found');
      return res.json(task);
    } catch (err: any) {
      console.error('Error in getTaskById:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async createTask(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const creatorId = req.user.id;
      const { title, description, status, priority, dueDate, estimatedTime, projectId, assigneeId, parentId } = req.body;

      const resolvedProjectId = projectId || (req.body.project ? req.body.project.id : null);
      if (!resolvedProjectId) {
        return res.status(400).send('Project ID is required');
      }

      const project = await Project.findByPk(resolvedProjectId);
      if (!project) return res.status(404).send('Project not found');

      let taskPriority = priority;
      if (!taskPriority) {
        taskPriority = AIService.predictPriority(title, description);
      }

      let taskEstTime = estimatedTime;
      const resolvedAssigneeId = assigneeId || (req.body.assignee ? req.body.assignee.id : null);
      if (!taskEstTime || Number(taskEstTime) === 0) {
        let experience = 2;
        if (resolvedAssigneeId) {
          const assigneeUser = await User.findByPk(resolvedAssigneeId);
          if (assigneeUser) experience = assigneeUser.experience || 2;
        }
        taskEstTime = AIService.estimateDuration(title, description, experience);
      }

      let taskStatus = status || 'TO_DO';
      if (resolvedAssigneeId) {
        const assigneeUser = await User.findByPk(resolvedAssigneeId);
        if (assigneeUser && assigneeUser.role === Role.ROLE_EMPLOYEE && taskStatus !== 'COMPLETED') {
          taskStatus = 'PENDING_ACCEPTANCE';
        }
      }

      const task = await Task.create({
        title: title || 'New Task',
        description,
        status: taskStatus,
        priority: taskPriority,
        dueDate,
        estimatedTime: Number(taskEstTime || 0),
        actualTime: 0.0,
        projectId: resolvedProjectId,
        assigneeId: resolvedAssigneeId || null,
        creatorId,
        parentTaskId: parentId ? Number(parentId) : null,
      });

      if (resolvedAssigneeId) {
        const assigneeUser = await User.findByPk(resolvedAssigneeId);
        const creatorUser = await User.findByPk(creatorId);

        if (assigneeUser) {
          await Notification.create({
            title: 'Task Assigned',
            message: `You have been assigned: ${task.title}`,
            type: 'TASK_ASSIGNED',
            isRead: false,
            recipientId: assigneeUser.id,
          });

          await DirectMessage.create({
            content: `Hello, I have assigned you a new task: "${task.title}". Please review and accept it.`,
            senderId: creatorId,
            recipientId: assigneeUser.id,
            taskId: task.id,
            isRead: false,
          });
        }
      }

      const savedTask = await Task.findByPk(task.id, {
        include: [
          { model: Project, as: 'project' },
          { model: User, as: 'assignee', attributes: { exclude: ['password'] } },
          { model: User, as: 'creator', attributes: { exclude: ['password'] } },
        ],
      });

      return res.status(201).json(savedTask);
    } catch (err: any) {
      console.error('Error in createTask:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async updateTask(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const id = parseInt(req.params.id);
      const task = await Task.findByPk(id);
      if (!task) return res.status(404).send('Task not found');

      const currentUser = await User.findByPk(req.user.id);
      const isTeamLeader = req.user.role === Role.ROLE_ADMIN || req.user.role === Role.ROLE_MANAGER;

      if (!isTeamLeader && task.assigneeId !== req.user.id) {
        return res.status(403).send('Only the assigned member or Team Leader can update this task.');
      }

      const oldStatus = task.status;
      const oldAssigneeId = task.assigneeId;
      const incomingStatus = req.body.status;

      // Duplicate acceptance guard
      if ((incomingStatus === 'ACCEPTED' || incomingStatus === 'TO_DO') && oldStatus === 'ACCEPTED') {
        const assigneeObj = oldAssigneeId ? await User.findByPk(oldAssigneeId) : null;
        return res.status(400).send(`Task has already been accepted by ${assigneeObj ? assigneeObj.name : 'the assignee'}.`);
      }

      const { title, description, priority, dueDate, estimatedTime, actualTime, recurring, recurringPattern, declineReason, assignee, assigneeId } = req.body;

      if (title !== undefined) task.title = title;
      if (description !== undefined) task.description = description;
      if (incomingStatus !== undefined) task.status = incomingStatus;
      if (priority !== undefined) task.priority = priority;
      if (dueDate !== undefined) task.dueDate = dueDate;
      if (estimatedTime !== undefined) task.estimatedTime = Number(estimatedTime);
      if (actualTime !== undefined) task.actualTime = Number(actualTime);
      if (recurring !== undefined) task.recurring = recurring;
      if (recurringPattern !== undefined) task.recurringPattern = recurringPattern;
      if (declineReason !== undefined) task.declineReason = declineReason;

      if (incomingStatus === 'ACCEPTED' && oldStatus === 'PENDING_ACCEPTANCE') {
        task.acceptedAt = new Date();
      }

      const resolvedAssigneeId = assigneeId !== undefined ? assigneeId : (assignee ? assignee.id : undefined);

      if (resolvedAssigneeId !== undefined) {
        if (resolvedAssigneeId === null) {
          task.assigneeId = null;
        } else {
          const newAssignee = await User.findByPk(resolvedAssigneeId);
          if (newAssignee) {
            task.assigneeId = newAssignee.id;
            if (oldAssigneeId !== newAssignee.id) {
              if (newAssignee.role === Role.ROLE_EMPLOYEE && incomingStatus !== 'COMPLETED') {
                task.status = 'PENDING_ACCEPTANCE';
              }
              await Notification.create({
                title: 'Task Assigned',
                message: `You have been assigned: ${task.title}`,
                type: 'TASK_ASSIGNED',
                isRead: false,
                recipientId: newAssignee.id,
              });

              await DirectMessage.create({
                content: `Hello, I have assigned you a new task: "${task.title}". Please review and accept it.`,
                senderId: req.user.id,
                recipientId: newAssignee.id,
                taskId: task.id,
                isRead: false,
              });
            }
          }
        }
      }

      await task.save();

      // Notifications on state transition
      if (oldStatus === 'PENDING_ACCEPTANCE' && (task.status === 'ACCEPTED' || task.status === 'TO_DO') && oldAssigneeId) {
        // Auto-dismiss TASK_ASSIGNED notifications
        await Notification.update({ isRead: true }, {
          where: {
            recipientId: oldAssigneeId,
            type: 'TASK_ASSIGNED',
            message: { [Op.like]: `%${task.title}%` },
          },
        });

        const oldAssigneeObj = await User.findByPk(oldAssigneeId);
        await DirectMessage.create({
          content: `I have accepted the task: "${task.title}".`,
          senderId: oldAssigneeId,
          recipientId: task.creatorId,
          taskId: task.id,
          isRead: false,
        });

        await Notification.create({
          title: 'Task Accepted',
          message: `${oldAssigneeObj?.name || 'Assignee'} accepted: "${task.title}"`,
          type: 'TASK_ACCEPTED',
          isRead: false,
          recipientId: task.creatorId,
        });
      }

      if (oldStatus === 'PENDING_ACCEPTANCE' && (task.status === 'BACKLOG' || task.status === 'DECLINED') && oldAssigneeId) {
        const oldAssigneeObj = await User.findByPk(oldAssigneeId);
        let reason = declineReason || 'No reason provided.';
        if (!reason.startsWith(oldAssigneeObj?.name || '')) {
          reason = `${oldAssigneeObj?.name || 'Assignee'}: ${reason}`;
        }
        task.declineReason = reason;
        await task.save();

        await DirectMessage.create({
          content: `I have declined the task: "${task.title}". Reason: ${reason}`,
          senderId: oldAssigneeId,
          recipientId: task.creatorId,
          taskId: task.id,
          isRead: false,
        });

        await Notification.create({
          title: 'Task Declined',
          message: `${oldAssigneeObj?.name || 'Assignee'} declined: "${task.title}". Reason: ${reason}`,
          type: 'TASK_DECLINED',
          isRead: false,
          recipientId: task.creatorId,
        });
      }

      if (task.status === 'COMPLETED' && oldStatus !== 'COMPLETED') {
        await Notification.create({
          title: 'Task Completed',
          message: `The task: ${task.title} has been marked complete.`,
          type: 'TASK_COMPLETED',
          isRead: false,
          recipientId: task.creatorId,
        });
      }

      const updatedTask = await Task.findByPk(id, {
        include: [
          { model: Project, as: 'project' },
          { model: User, as: 'assignee', attributes: { exclude: ['password'] } },
          { model: User, as: 'creator', attributes: { exclude: ['password'] } },
        ],
      });

      return res.json(updatedTask);
    } catch (err: any) {
      console.error('Error in updateTask:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async deleteTask(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const id = parseInt(req.params.id);
      const task = await Task.findByPk(id);
      if (!task) return res.status(404).send('Task not found');

      if (req.user.role !== Role.ROLE_ADMIN && req.user.role !== Role.ROLE_MANAGER) {
        return res.status(403).send('Only Team Leaders can delete tasks.');
      }

      await task.destroy();
      return res.json({ success: true });
    } catch (err: any) {
      console.error('Error in deleteTask:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async getComments(req: AuthRequest, res: Response) {
    try {
      const taskId = parseInt(req.params.id);
      const comments = await Comment.findAll({
        where: { taskId },
        include: [{ model: User, as: 'user', attributes: { exclude: ['password'] } }],
        order: [['createdAt', 'ASC']],
      });
      return res.json(comments);
    } catch (err: any) {
      console.error('Error in getComments:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async addComment(req: AuthRequest, res: Response) {
    try {
      if (!req.user) return res.status(401).send('Unauthorized');

      const taskId = parseInt(req.params.id);
      const { content } = req.body;

      const task = await Task.findByPk(taskId);
      if (!task) return res.status(404).send('Task not found');

      const comment = await Comment.create({
        content: content || '',
        taskId,
        userId: req.user.id,
      });

      if (task.assigneeId && task.assigneeId !== req.user.id) {
        await Notification.create({
          title: 'New Comment',
          message: `${req.user.name || 'User'} commented on: ${task.title}`,
          type: 'MENTION',
          isRead: false,
          recipientId: task.assigneeId,
        });
      }

      const fullComment = await Comment.findByPk(comment.id, {
        include: [{ model: User, as: 'user', attributes: { exclude: ['password'] } }],
      });

      return res.json(fullComment);
    } catch (err: any) {
      console.error('Error in addComment:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async updateTimer(req: AuthRequest, res: Response) {
    try {
      const taskId = parseInt(req.params.id);
      const additionalHours = parseFloat(String(req.query.additionalHours || req.body.additionalHours || 0));

      const task = await Task.findByPk(taskId);
      if (!task) return res.status(404).send('Task not found');

      task.actualTime = (task.actualTime || 0) + additionalHours;
      await task.save();

      const project = await Project.findByPk(task.projectId);
      if (project) {
        project.spent = (project.spent || 0) + additionalHours * 50.0;
        await project.save();
      }

      return res.json(task);
    } catch (err: any) {
      console.error('Error in updateTimer:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async getAiSubtasks(req: AuthRequest, res: Response) {
    try {
      const taskId = parseInt(req.params.id);
      const task = await Task.findByPk(taskId);
      if (!task) return res.status(404).send('Task not found');

      const suggestions = AIService.suggestSubtasks(task.title, task.description);
      return res.json(suggestions);
    } catch (err: any) {
      console.error('Error in getAiSubtasks:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async addDependency(req: AuthRequest, res: Response) {
    try {
      const id = parseInt(req.params.id);
      const depId = parseInt(req.params.depId);

      const task = await Task.findByPk(id);
      const dependency = await Task.findByPk(depId);

      if (!task || !dependency) return res.status(404).send('Task or Dependency not found');

      await (task as any).addDependency(dependency);

      const updatedTask = await Task.findByPk(id, {
        include: [{ model: Task, as: 'dependencies', through: { attributes: [] } }],
      });

      return res.json(updatedTask);
    } catch (err: any) {
      console.error('Error in addDependency:', err);
      return res.status(500).json({ error: err.message });
    }
  }

  public static async removeDependency(req: AuthRequest, res: Response) {
    try {
      const id = parseInt(req.params.id);
      const depId = parseInt(req.params.depId);

      const task = await Task.findByPk(id);
      const dependency = await Task.findByPk(depId);

      if (!task || !dependency) return res.status(404).send('Task or Dependency not found');

      await (task as any).removeDependency(dependency);

      const updatedTask = await Task.findByPk(id, {
        include: [{ model: Task, as: 'dependencies', through: { attributes: [] } }],
      });

      return res.json(updatedTask);
    } catch (err: any) {
      console.error('Error in removeDependency:', err);
      return res.status(500).json({ error: err.message });
    }
  }
}
